gcp/appengine/generate_sitemap.py

```diff
@@ -0,0 +1,122 @@
+# Copyright 2021 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Generate sitemap."""
+import sys
+import os
+import osv
+import osv.logs
+import datetime
+import argparse
+from google.cloud import ndb
+
+from xml.etree.ElementTree import Element, SubElement, ElementTree
+
+_SITEMAPS_DIRECTORY = './sitemap'
+_SITEMAP_INDEX_PATH = f'{_SITEMAPS_DIRECTORY}/index.xml'
+
+
+def fetch_vulnerability_ids(ecosystem: str) -> list[str]:
+  """Fetch vulnerabilities' id for the given ecosystem."""
+  bugs = osv.Bug.query(
+      osv.Bug.status == osv.BugStatus.PROCESSED,
+      osv.Bug.public == True,  # pylint: disable=singleton-comparison
+      osv.Bug.ecosystem == ecosystem).order(-osv.Bug.timestamp)
+  bug_ids = [bug.db_id for bug in bugs]
+  return bug_ids
+
+
+def osv_get_ecosystems():
+  """Get list of ecosystems."""
+  query = osv.Bug.query(projection=[osv.Bug.ecosystem], distinct=True)
+  return sorted([bug.ecosystem[0] for bug in query if bug.ecosystem],
+                key=str.lower)
+
+
+def get_sitemap_filename_for_ecosystem(ecosystem: str) -> str:
+  ecosystem_name = ecosystem.replace(' ', '_').replace('.', '__').strip()
+  return f'{_SITEMAPS_DIRECTORY}/{ecosystem_name}.xml'
+
+
+def get_sitemap_url_for_ecosystem(ecosystem: str, base_url: str) -> str:
+  ecosystem_name = ecosystem.replace(' ', '_').replace('.', '__').strip()
+  return f'{base_url}/sitemap/{ecosystem_name}.xml'
+
+
+def generate_sitemap_for_ecosystem(ecosystem: str, base_url: str) -> None:
+  """Generate a sitemap for the give n ecosystem."""
+  os.makedirs(_SITEMAPS_DIRECTORY, exist_ok=True)
+
+  vulnerability_ids = fetch_vulnerability_ids(ecosystem)
+  filename = get_sitemap_filename_for_ecosystem(ecosystem)
+  urlset = Element(
+      'urlset', xmlns="http://www.sitemaps.org/schemas/sitemap/0.9")
+
+  for vuln in vulnerability_ids:
```

Comment:
Limit this to 50000 (or 49999) to avoid generating an invalid sitemap.

Reply:
updated!